packages/react-i18n-webpack-plugin/src/react-i18n-webpack-plugin.ts

```diff
@@ -0,0 +1,264 @@
+import path from 'path';
+
+import webpack from 'webpack';
+import {camelCase} from 'change-case';
+import VirtualModulesPlugin from 'webpack-virtual-modules';
+import ParserHelpers from 'webpack/lib/ParserHelpers';
+import {CallExpression, Expression, SpreadElement} from 'estree';
+
+import {FindI18nImportPlugin} from './find-i18n-imports-plugin';
+import {generateID} from './utilities';
+
+const PLUGIN_NAME = 'ReactI18nPlugin';
+const TRANSLATION_DIRECTORY_NAME = 'translations';
+
+export interface Options {
+  fallbackLocale: string;
+}
+
+export class ReactI18nPlugin {
+  private options: Options;
+
+  private defaultOptions: Options = {
+    fallbackLocale: 'en',
+  };
+
+  constructor(options: Partial<Options> = {}) {
+    this.options = {
+      ...this.defaultOptions,
+      ...options,
+    };
+  }
+
+  apply(compiler: webpack.Compiler) {
+    const virtualModules = new VirtualModulesPlugin();
+    virtualModules.apply(compiler);
+
+    compiler.hooks.compilation.tap(
+      PLUGIN_NAME,
+      (
+        compilation: webpack.compilation.Compilation,
+        {
+          normalModuleFactory,
+        }: {normalModuleFactory: webpack.compilation.NormalModuleFactory},
+      ) => {
+        const handler = (parser: any) => {
+          new FindI18nImportPlugin().apply(parser);
+
+          // replace useI18n & withI18n call arguments
+          parser.hooks.evaluate
+            .for('CallExpression')
+            .tap(PLUGIN_NAME, (originalExpression: CallExpression) => {
+              if (
+                parser.state.module.resource.indexOf('node_modules') !== -1 &&
+                !parser.state.module.resource.endsWith('tsx') &&
+                !parser.state.i18nImports
+              ) {
+                return;
+              }
+
+              const componentPath = parser.state.module.resource;
+              const componentDir = parser.state.module.context;
+              const identifierName = parser.state.i18nImports.get(
+                componentPath,
+              );
+
+              if (
+                !identifierName ||
+                originalExpression.callee.type !== 'Identifier'
+              ) {
+                return;
+              }
+
+              let expression: CallExpression | null = null;
+              if (originalExpression.callee.name === 'compose') {
```

Comment:
I feel like we should never have to care if `compose` is called or not. Is this hook not called for every call expression?

Reply:
It does not. In the case of `compose`, this hook is called once with compose with `withI18n` as argument but no again with `withI18n`.